drift-client/src/main/java/io/airlift/drift/client/DriftMethodInvocation.java

```diff
@@ -0,0 +1,287 @@
+            AddressSelector addressSelector,
+            Optional<String> addressSelectionContext,
+            MethodInvocationStat stat,
+            Ticker ticker)
+    {
+        this.invoker = requireNonNull(invoker, "methodHandler is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.headers = requireNonNull(headers, "headers is null");
+        this.parameters = requireNonNull(parameters, "parameters is null");
+        this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
+        this.addressSelector = requireNonNull(addressSelector, "addressSelector is null");
+        this.addressSelectionContext = requireNonNull(addressSelectionContext, "addressSelectionContext is null");
+        this.stat = requireNonNull(stat, "stat is null");
+        this.ticker = requireNonNull(ticker, "ticker is null");
+        this.startTime = ticker.read();
+
+        // if this invocation is canceled, cancel the tasks
+        super.addListener(() -> {
+            if (super.isCancelled()) {
+                onCancel(wasInterrupted());
+            }
+        }, directExecutor());
+    }
+
+    private synchronized void nextAttempt()
+    {
+        try {
+            // request was already canceled
+            if (isCancelled()) {
+                return;
+            }
+
+            Optional<HostAndPort> address = addressSelector.selectAddress(addressSelectionContext);
+            if (!address.isPresent()) {
+                fail();
+                return;
+            }
+
+            if (invocationAttempts > 0) {
+                stat.recordRetry();
+            }
+
+            connectionAttempts++;
+
+            long invocationStartTime = ticker.read();
+            ListenableFuture<Object> result = invoker.invoke(new InvokeRequest(metadata, address.get(), headers, parameters));
+            stat.recordResult(invocationStartTime, result);
+            currentTask = result;
+
+            Futures.addCallback(result, new FutureCallback<Object>()
+            {
+                @Override
+                public void onSuccess(Object result)
+                {
+                    set(result);
+                }
+
+                @Override
+                public void onFailure(Throwable t)
+                {
+                    handleFailure(address.get(), t);
+                }
+            });
+        }
+        catch (Throwable t) {
+            // this should never happen, but ensure that invocation always finishes
+            unexpectedError(t);
+        }
+    }
+
+    private synchronized void handleFailure(HostAndPort address, Throwable throwable)
+    {
+        try {
+            ExceptionClassification exceptionClassification = retryPolicy.classifyException(throwable);
+
+            // update stats based on classification
+            if (exceptionClassification.getHostStatus() == NORMAL) {
+                // only store exception if the server was in a normal state
```

Comment:
Should this say _"server is in"_ instead of _"server was in"_? It's confusing because "was" implies previous state, but this is comparing the current state